rest_server/SiodbLogger.go

```diff
@@ -0,0 +1,186 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	TRACE   = 1
+	DEBUG   = 2
+	INFO    = 3
+	WARNING = 4
+	ERROR   = 5
+	FATAL   = 6
+)
+
+func ttos(t int) string {
+	switch t {
+	case 1:
+		return "trace"
+	case 2:
+		return "debug"
+	case 3:
+		return "info"
+	case 4:
+		return "warning"
+	case 5:
+		return "error"
+	case 6:
+		return "fatal"
+	default:
+		return "unknown"
+	}
+}
+
+type SiodbLoggerPool struct {
+	siodbLogger []SiodbLogger
+}
+
+func (l *SiodbLoggerPool) ConfigGinLogger() error {
+
+	file, err := os.Open("/dev/null")
+	if err != nil {
+		log.Fatal(err)
+	}
+	gin.DefaultWriter = file
+	gin.DisableConsoleColor()
+	gin.SetMode(gin.ReleaseMode)
+
+	return nil
+
+}
+
+func (l *SiodbLoggerPool) GinFormattedOutput(param gin.LogFormatterParams) string {
+
+	return FormattedOutput(INFO, "%s %s %s %s %d %s %s %s",
+		param.ClientIP,
+		param.Method,
+		param.Path,
+		param.Request.Proto,
+		param.StatusCode,
+		param.Latency,
+		param.Request.UserAgent(),
+		param.ErrorMessage,
+	)
+}
+
+func (l *SiodbLoggerPool) AddSiodbLogger(userLogLevel string, userLogDestination string) error {
+
+	// Derive logLevel
+	var logLevel int
+	switch strings.TrimSpace(strings.ToUpper(userLogLevel)) {
+	case "TRACE":
+		logLevel = TRACE
+	case "DEBUG":
+		logLevel = DEBUG
+	case "INFO":
+		logLevel = INFO
+	case "WARNING":
+		logLevel = WARNING
+	case "ERROR":
+		logLevel = ERROR
+	case "FATAL":
+		logLevel = FATAL
+	default:
+		log.Fatalf("Unknown log level: %s.\n",
+			strings.TrimSpace(strings.ToUpper(userLogLevel)))
+	}
+
+	// Derive log destination
+	var logDestination io.Writer
+	var err error
+	switch strings.TrimSpace(strings.ToLower(userLogDestination)) {
+	case "stdout":
+		logDestination = os.Stdout
+	case "stderr":
+		logDestination = os.Stderr
+	default: // Assuming it's a path
+
+		logFileName := fmt.Sprintf("rest_%v_%v.log",
+			time.Now().UTC().Format("20060102_150405"), unix.Getppid())
+		if _, err := os.Stat(strings.TrimSpace(strings.ToLower(userLogDestination))); os.IsNotExist(err) {
+			log.Fatalf("Unknown log destination: %s.\n",
+				strings.TrimSpace(strings.ToLower(userLogDestination)))
+		}
+		if logDestination, err = os.Create(
+			strings.TrimSpace(strings.ToLower(userLogDestination)) + "/" + logFileName); err != nil {
+			log.Fatalf("Can't create log file: %v.\n", err)
+		}
+
+	}
+
+	l.siodbLogger = append(l.siodbLogger,
+		SiodbLogger{
+			logLevel: logLevel,
+			out:      logDestination,
+		})
+
+	return nil
+
+}
+
+func (l *SiodbLoggerPool) Output(logLevel int, s string, v ...interface{}) error {
+
+	for _, siodbLogger := range l.siodbLogger {
+		siodbLogger.Output(logLevel, s, v...)
+	}
+
+	if logLevel == FATAL {
+		os.Exit(2)
```

Comment:
What if you run into fatal error after initialization?
It is important for us to distinguish these 2 situations - error during initialization or error after initialization - to control process properly from the `siodb` process. Exit codes MUST differ.

Reply:
Ok, with a new `FATALINIT` enum or with an additional parameter to the log function (`1` or `2` for initialization or afterinit)?

Reply:
First thing - this is absolutely not obvious that call to log with fatal level should cause program exit. I'd just put os.Exit with proper code after each log(fatal).

Reply:
So we need an additional parameter to `func (l *SiodbLoggerPool) Output(logLevel int, s string, v ...interface{}) error` to include the error code. 2 for init and >2 for others. Isn't it?

Reply:
No, you need to put `os.Exit(code)` after each fatal log output. Such code clearly states what is happening.